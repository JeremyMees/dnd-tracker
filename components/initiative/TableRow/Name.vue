<script setup lang="ts">
const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
}>()

const modal = useModal()

function openModal(): void {
  modal.open({
    component: 'InitiativeRowName',
    header: 'Name',
    props: {
      encounterId: props.sheet.id,
      name: props.item.name,
      submit: async (value: string) => {
        const index = getCurrentRowIndex(props.sheet, props.item.id)
        const rows = props.sheet.rows

        if (index === -1) return

        rows[index].name = value

        await props.update({ rows })
      },
    },
  })
}
</script>

<template>
  <td>
    <button
      class="flex items-center gap-x-2"
      @click="openModal"
    >
      <Icon
        :name="homebrewIcon(item.type)"
        :class="homebrewColor(item.type)"
        class="size-5 min-w-5"
        aria-hidden="true"
      />
      <div class="flex flex-col gap-y-1">
        <span>
          {{ item.name }}
        </span>
        <span
          v-if="item.summoner?.name"
          class="body-extra-small text-muted-foreground"
        >
          {{ $t('general.summoner') }}: {{ item.summoner.name }}
        </span>
      </div>
    </button>
  </td>
</template>
