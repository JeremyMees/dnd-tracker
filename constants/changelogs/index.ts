import v1_0_8 from './v1-0-8.json'
import v1_0_9 from './v1-0-9.json'
import v1_0_10 from './v1-0-10.json'
import v1_0_11 from './v1-0-11.json'
import v1_0_12 from './v1-0-12.json'
import v1_0_13 from './v1-0-13.json'
import v1_0_14 from './v1-0-14.json'
import v1_0_15 from './v1-0-15.json'
import v1_0_16 from './v1-0-16.json'
import v1_0_17 from './v1-0-17.json'
import v1_0_18 from './v1-0-18.json'
import v1_0_19 from './v1-0-19.json'
import v1_0_20 from './v1-0-20.json'
import v2_0_0 from './v2-0-0.json'

export const changelogs = [
  v2_0_0,
  v1_0_20,
  v1_0_19,
  v1_0_18,
  v1_0_17,
  v1_0_16,
  v1_0_15,
  v1_0_14,
  v1_0_13,
  v1_0_12,
  v1_0_11,
  v1_0_10,
  v1_0_9,
  v1_0_8,
]
