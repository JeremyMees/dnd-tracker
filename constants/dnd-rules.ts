export const crOptions = [
  { label: '0', value: 0.0 },
  { label: '1/8', value: 0.125 },
  { label: '1/4', value: 0.25 },
  { label: '1/2', value: 0.5 },
  { label: '1', value: 1.0 },
  { label: '2', value: 2.0 },
  { label: '3', value: 3.0 },
  { label: '4', value: 4.0 },
  { label: '5', value: 5.0 },
  { label: '6', value: 6.0 },
  { label: '7', value: 7.0 },
  { label: '8', value: 8.0 },
  { label: '9', value: 9.0 },
  { label: '10', value: 10.0 },
  { label: '11', value: 11.0 },
  { label: '12', value: 12.0 },
  { label: '13', value: 13.0 },
  { label: '14', value: 14.0 },
  { label: '15', value: 15.0 },
  { label: '16', value: 16.0 },
  { label: '17', value: 17.0 },
  { label: '18', value: 18.0 },
  { label: '19', value: 19.0 },
  { label: '20', value: 20.0 },
  { label: '21', value: 21.0 },
  { label: '22', value: 22.0 },
  { label: '23', value: 23.0 },
  { label: '24', value: 24.0 },
  { label: '30', value: 30.0 },
]

export const abilities = ['STR', 'DEX', 'CON', 'INT', 'WIS', 'CHA']

export const initiativeKeys = [
  'ac',
  'campaign',
  'concentration',
  'conditions',
  'created_at',
  'deathSaves',
  'health',
  'id',
  'index',
  'initiative',
  'initiative_modifier',
  'link',
  'maxAc',
  'maxHealth',
  'maxAcOld',
  'maxHealthOld',
  'name',
  'note',
  'tempAc',
  'tempHealth',
  'type',
  'summoner',
  'actions',
  'legendary_actions',
  'reactions',
  'special_abilities',
]
