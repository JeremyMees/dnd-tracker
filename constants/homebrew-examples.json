[
  {
    "type": "player",
    "id": 1,
    "name": "Carlo",
    "player": "Thomas",
    "health": 37,
    "ac": 13,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "player",
    "id": 2,
    "name": "Silvin",
    "player": "Sander",
    "health": 33,
    "ac": 14,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "player",
    "id": 3,
    "name": "Alexis",
    "player": "Davina",
    "health": 29,
    "ac": 12,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "player",
    "id": 4,
    "name": "Diadematus",
    "player": "Tom",
    "health": 31,
    "ac": 16,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "player",
    "id": 5,
    "name": "Banthaaja",
    "player": "Yves",
    "health": 30,
    "ac": 15,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "monster",
    "id": 6,
    "name": "Lulu",
    "health": 69,
    "ac": 20,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "monster",
    "id": 7,
    "name": "Pony",
    "health": 33,
    "ac": 10,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "npc",
    "id": 8,
    "name": "Rocky",
    "health": 24,
    "ac": 16,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "npc",
    "id": 9,
    "name": "Gray",
    "health": "20",
    "ac": "16",
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "npc",
    "id": 10,
    "name": "Ginger",
    "health": 10,
    "ac": 18,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "summon",
    "id": 11,
    "name": "Inli",
    "health": 18,
    "ac": 12,
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  },
  {
    "type": "lair",
    "id": 12,
    "name": "Lair action",
    "created_at": "now",
    "campaign": 1,
    "actions": [],
    "legendary_actions": [],
    "reactions": [],
    "special_abilities": [],
    "initiative_modifier": null,
    "link": null,
    "source": null
  }
]