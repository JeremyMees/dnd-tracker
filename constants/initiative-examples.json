{
  "id": 2,
  "created_at": "2022-12-07T07:34:18+00:00",
  "created_by": "me :)",
  "title": "Sandbox",
  "round": 1,
  "activeIndex": 0,
  "rows": [
    {
      "id": "0",
      "name": "Lair action",
      "type": "lair",
      "index": 0,
      "conditions": [],
      "initiative": 20,
      "initiative_modifier": 2,
      "link": null
    },
    {
      "id": "1",
      "name": "Kees",
      "type": "player",
      "index": 1,
      "health": 69,
      "maxHealth": 69,
      "ac": 22,
      "maxAc": 22,
      "tempAc": 0,
      "tempHealth": 0,
      "conditions": [],
      "initiative": 19,
      "initiative_modifier": 1,
      "concentration": true,
      "link": "https://www.youtube.com/watch?v=uxauCAtFckw",
      "deathSaves": {
        "fail": [
          false,
          false,
          false
        ],
        "save": [
          false,
          false,
          false
        ]
      }
    },
    {
      "id": 2,
      "name": "Nuxt",
      "type": "player",
      "index": 2,
      "health": 0,
      "maxHealth": 60,
      "ac": 15,
      "maxAc": 15,
      "tempAc": 0,
      "tempHealth": 0,
      "conditions": [
        {
          "slug": "exhaustion",
          "name": "Exhaustion",
          "desc": "* Some special abilities and environmental hazards, such as starvation and the long-term effects of freezing or scorching temperatures, can lead to a special condition called exhaustion. Exhaustion is measured in six levels. An effect can give a creature one or more levels of exhaustion, as specified in the effect's description.\n\n| Level | Effect                                         |\n|-------|------------------------------------------------|\n| 1     | Disadvantage on ability checks                 |\n| 2     | Speed halved                                   |\n| 3     | Disadvantage on attack rolls and saving throws |\n| 4     | Hit point maximum halved                       |\n| 5     | Speed reduced to 0                             |\n| 6     | Death                                          |\n\nIf an already exhausted creature suffers another effect that causes exhaustion, its current level of exhaustion increases by the amount specified in the effect's description.\n\nA creature suffers the effect of its current level of exhaustion as well as all lower levels. For example, a creature suffering level 2 exhaustion has its speed halved and has disadvantage on ability checks.\n\nAn effect that removes exhaustion reduces its level as specified in the effect's description, with all exhaustion effects ending if a creature's exhaustion level is reduced below 1.\n\nFinishing a long rest reduces a creature's exhaustion level by 1, provided that the creature has also ingested some food and drink.",
          "document__slug": "wotc-srd",
          "document__title": "5e Core Rules",
          "document__url": "http://dnd.wizards.com/articles/features/systems-reference-document-srd",
          "hasLevels": true,
          "level": 3
        }
      ],
      "initiative": 19,
      "initiative_modifier": 4,
      "concentration": false,
      "link": "https://www.youtube.com/embed/y90yaLFoYoA?si=zWL9zZslBib6Fcni&amp;controls=0",
      "deathSaves": {
        "fail": [
          true,
          true,
          false
        ],
        "save": [
          true,
          false,
          false
        ]
      }
    },
    {
      "id": 3,
      "name": "Billy Bob",
      "type": "player",
      "index": 3,
      "health": 70,
      "maxHealth": 70,
      "ac": 16,
      "maxAc": 16,
      "tempAc": 0,
      "tempHealth": 0,
      "conditions": [],
      "initiative": 19,
      "initiative_modifier": 3,
      "concentration": false,
      "link": "https://www.youtube.com/embed/AWM5ZNdWlqw?si=l-87E9B2t65gR35X&amp;controls=0",
      "deathSaves": {
        "fail": [
          false,
          false,
          false
        ],
        "save": [
          false,
          false,
          false
        ]
      }
    },
    {
      "id": 1682257367892,
      "name": "Dragon Turtle",
      "type": "monster",
      "index": 4,
      "health": 341,
      "maxHealth": 341,
      "ac": 20,
      "maxAc": 20,
      "tempAc": 0,
      "tempHealth": 0,
      "conditions": [],
      "initiative": 17,
      "initiative_modifier": 2,
      "concentration": false,
      "link": "https://www.aidedd.org/dnd/monstres.php?vo=dragon-turtle",
      "note": "",
      "deathSaves": {
        "fail": [
          false,
          false,
          false
        ],
        "save": [
          false,
          false,
          false
        ]
      },
      "actions": [
        {
          "name": "Multiattack",
          "desc": "The dragon turtle makes three attacks: one with its bite and two with its claws. It can make one tail attack in place of its two claw attacks.",
          "type": "actions"
        },
        {
          "name": "Bite",
          "desc": "Melee Weapon Attack: +13 to hit, reach 15 ft., one target. Hit: 26 (3d12 + 7) piercing damage.",
          "attack_bonus": 13,
          "damage_dice": "3d12",
          "damage_bonus": 7,
          "type": "actions"
        },
        {
          "name": "Claw",
          "desc": "Melee Weapon Attack: +13 to hit, reach 10 ft., one target. Hit: 16 (2d8 + 7) slashing damage.",
          "attack_bonus": 13,
          "damage_dice": "2d8",
          "damage_bonus": 7,
          "type": "actions"
        },
        {
          "name": "Tail",
          "desc": "Melee Weapon Attack: +13 to hit, reach 15 ft., one target. Hit: 26 (3d12 + 7) bludgeoning damage. If the target is a creature, it must succeed on a DC 20 Strength saving throw or be pushed up to 10 feet away from the dragon turtle and knocked prone.",
          "attack_bonus": 13,
          "damage_dice": "3d12",
          "damage_bonus": 7,
          "type": "actions"
        },
        {
          "name": "Steam Breath (Recharge 5-6)",
          "desc": "The dragon turtle exhales scalding steam in a 60-foot cone. Each creature in that area must make a DC 18 Constitution saving throw, taking 52 (15d6) fire damage on a failed save, or half as much damage on a successful one. Being underwater doesn't grant resistance against this damage.",
          "attack_bonus": 0,
          "damage_dice": "15d6",
          "type": "actions"
        }
      ],
      "special_abilities": [
        {
          "name": "Amphibious",
          "desc": "The dragon turtle can breathe air and water.",
          "type": "special_abilities"
        }
      ]
    },
    {
      "id": 5,
      "name": "Beast",
      "type": "summon",
      "index": 5,
      "health": 20,
      "maxHealth": 20,
      "ac": 12,
      "maxAc": 12,
      "tempAc": 0,
      "tempHealth": 0,
      "summoner": {
        "name": "Kees",
        "id": "3"
      },
      "conditions": [],
      "initiative": 16,
      "initiative_modifier": 2,
      "concentration": false,
      "link": null
    },
    {
      "id": 1682257384816,
      "name": "Water Elemental",
      "type": "monster",
      "index": 6,
      "health": 114,
      "maxHealth": 114,
      "ac": 14,
      "maxAc": 14,
      "tempAc": 0,
      "tempHealth": 0,
      "conditions": [],
      "initiative": 7,
      "initiative_modifier": 6,
      "concentration": false,
      "link": "https://www.aidedd.org/dnd/monstres.php?vo=water-elemental",
      "note": "",
      "deathSaves": {
        "fail": [
          false,
          false,
          false
        ],
        "save": [
          false,
          false,
          false
        ]
      },
      "actions": [
        {
          "name": "Multiattack",
          "desc": "The elemental makes two slam attacks.",
          "type": "actions"
        },
        {
          "name": "Slam",
          "desc": "Melee Weapon Attack: +7 to hit, reach 5 ft., one target. Hit: 13 (2d8 + 4) bludgeoning damage.",
          "attack_bonus": 7,
          "damage_dice": "2d8",
          "damage_bonus": 4,
          "type": "actions"
        },
        {
          "name": "Whelm (Recharge 4-6)",
          "desc": "Each creature in the elemental's space must make a DC 15 Strength saving throw. On a failure, a target takes 13 (2d8 + 4) bludgeoning damage. If it is Large or smaller, it is also grappled (escape DC 14). Until this grapple ends, the target is restrained and unable to breathe unless it can breathe water. If the saving throw is successful, the target is pushed out of the elemental's space.\nThe elemental can grapple one Large creature or up to two Medium or smaller creatures at one time. At the start of each of the elemental's turns, each target grappled by it takes 13 (2d8 + 4) bludgeoning damage. A creature within 5 feet of the elemental can pull a creature or object out of it by taking an action to make a DC 14 Strength and succeeding.",
          "type": "actions"
        }
      ],
      "special_abilities": [
        {
          "name": "Water Form",
          "desc": "The elemental can enter a hostile creature's space and stop there. It can move through a space as narrow as 1 inch wide without squeezing.",
          "type": "special_abilities"
        },
        {
          "name": "Freeze",
          "desc": "If the elemental takes cold damage, it partially freezes; its speed is reduced by 20 ft. until the end of its next turn.",
          "type": "special_abilities"
        }
      ]
    },
    {
      "id": 7,
      "name": "Random npc",
      "type": "npc",
      "index": 7,
      "health": 25,
      "maxHealth": 25,
      "ac": 10,
      "maxAc": 10,
      "tempAc": 0,
      "tempHealth": 0,
      "conditions": [
        {
          "slug": "charmed",
          "name": "Charmed",
          "desc": "* A charmed creature can't attack the charmer or target the charmer with harmful abilities or magical effects.\n* The charmer has advantage on any ability check to interact socially with the creature.",
          "document__slug": "wotc-srd",
          "document__title": "Systems Reference Document"
        }
      ],
      "initiative": 1,
      "initiative_modifier": -1,
      "concentration": false,
      "link": null,
      "deathSaves": {
        "fail": [
          false,
          false,
          false
        ],
        "save": [
          false,
          false,
          false
        ]
      }
    }
  ],
  "info": null,
  "info_cards": [],
  "settings": {
    "spacing": "normal",
    "pet": "cat",
    "rows": [
      "summoner",
      "ac",
      "health",
      "conditions",
      "note",
      "deathSaves",
      "concentration",
      "modify"
    ],
    "modified": false
  }
}