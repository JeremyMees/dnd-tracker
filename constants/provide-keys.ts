export const INITIATIVE_SHEET: InjectionKey<{
  sheet: Ref<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
  activeRow: Ref<InitiativeSheetRow | undefined>
}> = Symbol('initiativeSheet')
