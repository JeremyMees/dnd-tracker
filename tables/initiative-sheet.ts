import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
