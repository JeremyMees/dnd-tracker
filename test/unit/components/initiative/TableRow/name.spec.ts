import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import Name from '~/components/initiative/TableRow/Name.vue'
import { INITIATIVE_SHEET } from '~~/constants/provide-keys'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  item: InitiativeSheetRow
}

type MockFunctions = {
  handleSubmit: (
    form: { name: string },
    node: { clearErrors: () => void, setErrors: (error: string) => void }
  ) => Promise<void>
}

const mockUpdate = vi.fn()
const mockSheet = ref<InitiativeSheet>(sheet)

const mockNode = {
  clearErrors: vi.fn(),
  setErrors: vi.fn(),
}

const provide = {
  [INITIATIVE_SHEET]: {
    sheet: mockSheet,
    update: mockUpdate,
  },
}

const props: Props = {
  item: sheet.rows[0]!,
}

describe('Initiative table row name', async () => {
  beforeEach(() => {
    mockUpdate.mockClear()
    mockSheet.value = sheet
  })

  it('Should match snapshot', async () => {
    const component = await mountSuspended(Name, { props, provide })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should display name and summoner if available', async () => {
    const component = await mountSuspended(Name, {
      props: {
        item: {
          ...props.item,
          summoner: { name: 'Summoner', id: '123' },
        },
      },
      provide,
    })

    expect(component.find('[data-test-name]').text()).toBe(props.item.name)
    expect(component.find('[data-test-summoner]').text()).toBe('general.summoner: Summoner')
  })

  it('Should not display summoner', async () => {
    const component = await mountSuspended(Name, { props, provide })

    expect(component.find('[data-test-summoner]').exists()).toBeFalsy()
  })

  it('Should handle name update', async () => {
    const component = await mountSuspended(Name, { props, provide })
    const newName = 'New Character Name'

    const vm = component.vm as unknown as MockFunctions
    await vm.handleSubmit({ name: newName }, mockNode)

    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining([
        expect.objectContaining({
          ...props.item,
          name: newName,
        }),
      ]),
    })
  })

  it('Should handle update errors', async () => {
    const errorUpdate = vi.fn().mockRejectedValue(new Error('Update failed'))

    // Update the provide's update function for this test
    const errorProvide = {
      [INITIATIVE_SHEET]: {
        sheet: mockSheet,
        update: errorUpdate,
      },
    }

    const component = await mountSuspended(Name, {
      props,
      provide: errorProvide,
    })

    const vm = component.vm as unknown as MockFunctions
    await vm.handleSubmit({ name: 'New Name' }, mockNode)

    expect(mockNode.clearErrors).toHaveBeenCalled()
    expect(mockNode.setErrors).toHaveBeenCalled()
  })

  it('Should not update if sheet is undefined', async () => {
    mockSheet.value = undefined as any

    const component = await mountSuspended(Name, {
      props,
      provide,
    })

    const vm = component.vm as unknown as MockFunctions
    await vm.handleSubmit({ name: 'New Name' }, mockNode)

    expect(mockUpdate).not.toHaveBeenCalled()
  })
})
