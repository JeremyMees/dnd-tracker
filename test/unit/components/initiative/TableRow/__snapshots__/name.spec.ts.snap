// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row name > Should match snapshot 1`] = `
"<button id="reka-popover-trigger-v-0-0-0" type="button" aria-haspopup="dialog" aria-expanded="false" aria-controls="" data-state="closed" class="flex items-center gap-x-2"><span class="iconify i-tabler:sword text-primary size-5 min-w-5" aria-hidden="true"></span>
  <div class="flex flex-col gap-y-1 text-left"><span data-test-name="">Sister IarnÄƒ</span>
    <!--v-if-->
  </div>
</button>
<!--teleport start-->
<!--teleport end-->"
`;
